### PR TITLE
fix: rebuild project tree from scratch when listing tests

### LINKED ISSUE #30396: [Bug]: Missing projects/tests when using multiple projects (folders) from 1.43.x
### Version

1.43.1

### Steps to reproduce

1. Folder structure:

Screenshot 2024-04-17 at 13 21 37

2. Playwright config:

```
export default defineConfig({
  ...
  /* Configure projects for major browsers */
  projects: [
    // all tests except send money
    {
      name: 'setup',
      testDir: './utils',
      testMatch: /auth.setup\.ts/,
    },
    {
      name: 'chromium',
      use: { ...devices['Desktop Chrome'] },
      dependencies: ['setup'],
      testDir: './tests/all-tests',
    },
    {
      name: 'firefox',
      use: { ...devices['Desktop Firefox'] },
      dependencies: ['setup'],
      testDir: './tests/all-tests',
    },
    // send money tests
    {
      name: 'send-money-setup',
      testDir: './utils',
      testMatch: /send-money-auth.setup\.ts/,
    },
    {
      name: 'chromium',
      use: { ...devices['Desktop Chrome'] },
      dependencies: ['send-money-setup'],
      testDir: './tests/send-money-tests',
    },
    {
      name: 'firefox',
      use: { ...devices['Desktop Firefox'] },
      dependencies: ['send-money-setup'],
      testDir: './tests/send-money-tests',
    },
    ...
```    

### Expected behavior

In previous version (1.42.1) UI mode / tests runner has all tests grouped by projects:

Screenshot 2024-04-17 at 13 19 53

### Actual behavior

In latest version (1.43.1) UI mode / tests runner contains only one folder, set unique name for project doesn't change anything.

Screenshot 2024-04-17 at 13 18 18

### Additional context

_No response_

### Environment

```shell
System:
    OS: macOS 14.4.1
    CPU: (12) arm64 Apple M2 Pro
    Memory: 82.56 MB / 32.00 GB
  Binaries:
    Node: 19.7.0 - ~/.nvm/versions/node/v19.7.0/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v19.7.0/bin/yarn
    npm: 9.6.1 - ~/.nvm/versions/node/v19.7.0/bin/npm
    pnpm: 8.15.7 - /opt/homebrew/bin/pnpm
    bun: 1.0.19 - ~/.bun/bin/bun
  Languages:
    Bash: 3.2.57 - /bin/bash
```

### PR DESCRIPTION
**Pull request by yury-s**
Instead of filtering tests assuming there are no two projects with same name we always rebuild test tree from scratch and restore previos test results in the list mode.

Fixes https://github.com/microsoft/playwright/issues/30396

### CODE REVIEW

File: packages/trace-viewer/src/ui/teleSuiteUpdater.ts [hunk 1]
```diff
@@ -145,3 +149,36 @@ export class TeleSuiteUpdater {
...
+      }
+    }
+  }
+
+  restore(suite: TeleSuite) {
+    for (const entry of suite.entries()) {
+      if (entry.type === 'test') {
+        const results = this._testIdToResults.get(entry.id);
+        if (results)
+          entry.results = results;
```
Comment (dgozman) [inline packages/trace-viewer/src/ui/teleSuiteUpdater.ts:178]: This does not update `TeleTestCase._resultsMap`. Perhaps introduce something like `_addResult()`?

File: packages/playwright/src/isomorphic/teleReceiver.ts [hunk 2]
```diff
@@ -206,35 +211,6 @@ export class TeleReporterReceiver {
     projectSuite._project = this._parseProject(project);
     for (const suite of project.suites)
       this._mergeSuiteInto(suite, projectSuite);
-
-    // Remove deleted tests when listing. Empty suites will be auto-filtered
```
Comment (dgozman) [inline packages/playwright/src/isomorphic/teleReceiver.ts:210]: The test covering this code is failing!

**Review by dgozman — COMMENTED** [review]
The

File: packages/playwright/src/isomorphic/teleReceiver.ts (hunk 2 above)
Comment (yury-s) [reply packages/playwright/src/isomorphic/teleReceiver.ts:210]: Removed `this._options.onUpdate();` from onBegin/onEnd in testSuiteUpdater but I don't quite like it. Alternatively, we could restore the results in the onBegin of the reporter created in _createReporter.

Comment (yury-s) [reply packages/playwright/src/isomorphic/teleReceiver.ts:210]: Updated.

**Review by dgozman — APPROVED** [review]

**Pull request merged**